src/Compilers/Shared/DesktopBuildClient.cs

```diff
@@ -313,6 +317,41 @@ protected override int RunLocalCompilation(string[] arguments, BuildPaths buildP
             }
         }
 
+        // Protected for testing
+        protected static bool TryConnectToNamedPipeWithSpinWait(NamedPipeClientStream pipeStream,
+                                                                int timeoutMs,
+                                                                CancellationToken cancellationToken)
+        {
+            // .NET 4.5 implementation of NamedPipeStream.Connect busy waits the entire time.
+            // Work around is to spin wait.
+            const int maxWaitInterval = 50;
+            int elapsed = 0;
+            var sw = new SpinWait();
+            while (true)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                int waitTime = Math.Min(timeoutMs - elapsed, maxWaitInterval);
+                if (timeoutMs != Timeout.Infinite && waitTime <= 0)
+                {
+                    return false;
+                }
+
+                try
+                {
+                    pipeStream.Connect(waitTime);
+                    break;
+                }
+                catch (Exception e) when (e is IOException || e is TimeoutException)
+                {
+                    // Ignore timeout
+                }
+                elapsed += waitTime;
+                sw.SpinOnce();
```

Comment:
We are not accounting for how long this operation and other operations actually take. Consider capturing the current time on entry in whatever form is reliable and won't be affected by a change to a local clock on the machine. And then measure the time elapsed since that point in order to determine that it is time to give up.